src/Kestrel.Transport.Sockets/Internal/SocketAwaitableEventArgs.cs

```diff
@@ -5,74 +5,136 @@
+            }
+            else if (awaitableState != null)
+            {
+                throw new InvalidOperationException("Multiple continuations registered!");
+            }
 
-            _callback = null;
+        }
 
-            if (SocketError != SocketError.Success)
+        public int GetResult(short token)
+        {
+            SocketError error = SocketError;
+            int bytes = BytesTransferred;
+
+            if (error != SocketError.Success)
             {
-                ThrowSocketException(SocketError);
+                ThrowSocketException(error);
             }
 
-            return BytesTransferred;
+            return bytes;
 
             void ThrowSocketException(SocketError e)
             {
                 throw new SocketException((int)e);
             }
         }
 
-        public void OnCompleted(Action continuation)
+        public ValueTask<int> ReceiveAsync(Socket socket)
         {
-            if (ReferenceEquals(_callback, _callbackCompleted) ||
-                ReferenceEquals(Interlocked.CompareExchange(ref _callback, continuation, null), _callbackCompleted))
+            if (socket.ReceiveAsync(this))
             {
-                Task.Run(continuation);
+                return new ValueTask<int>(this, 0);
             }
+
+            return GetCompletedValueTask();
         }
 
-        public void UnsafeOnCompleted(Action continuation)
+        /// <summary>Initiates a send operation on the associated socket.</summary>
+        /// <returns>This instance.</returns>
+        public ValueTask<int> SendAsync(Socket socket)
         {
-            OnCompleted(continuation);
+            if (socket.SendAsync(this))
+            {
+                return new ValueTask<int>(this, 0);
+            }
+
+            return GetCompletedValueTask();
         }
 
-        public void Complete()
+        private ValueTask<int> GetCompletedValueTask()
         {
-            OnCompleted(this);
+            int bytesTransferred = BytesTransferred;
+            SocketError error = SocketError;
+
+            Volatile.Write(ref _callback, null);
+
+            return error == SocketError.Success ?
+                new ValueTask<int>(bytesTransferred) :
+                new ValueTask<int>(Task.FromException<int>(new SocketException((int)error)));
         }
 
-        protected override void OnCompleted(SocketAsyncEventArgs _)
+        private static void OnWorkItemCallback(object state)
         {
-            var continuation = Interlocked.Exchange(ref _callback, _callbackCompleted);
-
-            if (continuation != null)
-            {
-                _ioScheduler.Schedule(state => ((Action)state)(), continuation);
-            }
+            var args = (SocketAwaitableEventArgs)state;
+            var callbackState = args.UserToken;
+            args.UserToken = null;
+            var callback = args._callback;
+            callback(callbackState);
```

Comment:
I see what was up now. I always seem to forget that you can't convert an `Action<object>` to a `WaitCallback` without allocating.

Reply:
I'm going to change this back, it's a tad bit complicated and not worth saving the allocation.